js/ui/codingManager.js

```diff
@@ -29,239 +46,530 @@ function _isBuilderSpeedwagon(window) {
-    Name: 'CodingManager',
+function _createIcon() {
+    let iconFile = Gio.File.new_for_uri('resource:///org/gnome/shell/theme/rotate.svg');
+    let gicon = new Gio.FileIcon({ file: iconFile });
+    let icon = new St.Icon({ style_class: 'view-source-icon',
+                             gicon: gicon });
+    return icon;
+}
 
-    _init: function() {
-        this._sessions = [];
-        this._rotateInActors = [];
-        this._rotateOutActors = [];
-        this._watchdogId = 0;
-        this._codingApps = ['com.endlessm.Helloworld', 'org.gnome.Weather.Application'];
-    },
+function _getAppManifestAt(location, flatpakID) {
+    let manifestFile = Gio.File.new_for_path(GLib.build_filenamev([location, 'app', flatpakID, 'current',
+                                                                   'active', 'files', 'manifest.json']));
+    if (!manifestFile.query_exists(null))
+        return null;
+    return manifestFile;
+}
 
-    _isCodingApp: function(flatpakID) {
-        return this._codingApps.indexOf(flatpakID) != -1;
-    },
+function _getAppManifest(flatpakID) {
+    let manifestFile = _getAppManifestAt(Flatpak.Installation.new_user(null).get_path().get_path(), flatpakID);
+    if (manifestFile)
+        return manifestFile;
 
-    _isBuilder: function(flatpakID) {
-        return flatpakID === 'org.gnome.Builder';
+    manifestFile = _getAppManifestAt(Flatpak.Installation.new_system(null).get_path().get_path(), flatpakID);
+    if (manifestFile)
+        return manifestFile;
+
+    return null;
+}
+
+// _synchronizeMetaWindowActorGeometries
+//
+// Synchronize geometry of MetaWindowActor src to dst by
+// applying both the physical geometry and maximization state.
+function _synchronizeMetaWindowActorGeometries(src, dst) {
+    let srcGeometry = src.meta_window.get_frame_rect();
+    let dstGeometry = dst.meta_window.get_frame_rect();
+
+    let srcIsMaximized = (src.meta_window.maximized_horizontally &&
+                          src.meta_window.maximized_vertically);
+    let dstIsMaximized = (dst.meta_window.maximized_horizontally &&
+                          dst.meta_window.maximized_vertically);
+
+    if (!srcIsMaximized && dstIsMaximized)
+        dst.meta_window.unmaximize(Meta.MaximizeFlags.BOTH);
+
+    if (srcIsMaximized && !dstIsMaximized)
+        dst.meta_window.maximize(Meta.MaximizeFlags.BOTH);
+
+    if (srcGeometry.equal(dstGeometry))
+        return;
+
+    dst.meta_window.move_resize_frame(false,
+                                      srcGeometry.x,
+                                      srcGeometry.y,
+                                      srcGeometry.width,
+                                      srcGeometry.height);
+}
+
+function _synchronizeViewSourceButtonToRectCorner(button, rect) {
+    button.set_position(rect.x + rect.width - BUTTON_OFFSET_X,
+                        rect.y + rect.height - BUTTON_OFFSET_Y);
+}
+
+function _createViewSourceButtonInRectCorner(rect) {
+    let button = new St.Bin({
```

Comment:
I guess I was assuming this was an `St.Button`, not an `St.Bin` -- any reason not to make it such?